README.md
# Logstash Plugin

[![Travis Build Status](https://travis-ci.com/logstash-plugins/logstash-input-sqs.svg)](https://travis-ci.com/logstash-plugins/logstash-input-sqs)

This is a plugin for [Logstash](https://github.com/elastic/logstash).

It is fully free and fully open source. The license is Apache 2.0, meaning you are pretty much free to use it however you want in whatever way.

## Documentation

Logstash provides infrastructure to automatically generate documentation for this plugin. We use the asciidoc format to write documentation so any comments in the source code will be first converted into asciidoc and then into html. All plugin documentation are placed under one [central location](http://www.elastic.co/guide/en/logstash/current/).

- For formatting code or config example, you can use the asciidoc `[source,ruby]` directive
- For more asciidoc formatting tips, see the excellent reference here https://github.com/elastic/docs#asciidoc-guide

## Need Help?

Need help? Try #logstash on freenode IRC or the https://discuss.elastic.co/c/logstash discussion forum.

## Developing

### 1. Plugin Developement and Testing

#### Code
- To get started, you'll need JRuby with the Bundler gem installed.

- Create a new plugin or clone and existing from the GitHub [logstash-plugins](https://github.com/logstash-plugins) organization. We also provide [example plugins](https://github.com/logstash-plugins?query=example).

- Install dependencies
```sh
bundle install
```

#### Test

- Update your dependencies

```sh
bundle install
```

- Run tests

```sh
bundle exec rspec
```

### 2. Running your unpublished Plugin in Logstash

#### 2.1 Run in a local Logstash clone

- Edit Logstash `Gemfile` and add the local plugin path, for example:
```ruby
gem "logstash-filter-awesome", :path => "/your/local/logstash-filter-awesome"
```
- Install plugin
```sh
# Logstash 2.3 and higher
bin/logstash-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Run Logstash with your plugin
```sh
bin/logstash -e 'filter {awesome {}}'
```
At this point any modifications to the plugin code will be applied to this local Logstash setup. After modifying the plugin, simply rerun Logstash.

#### 2.2 Run in an installed Logstash

You can use the same **2.1** method to run your plugin in an installed Logstash by editing its `Gemfile` and pointing the `:path` to your local plugin development directory or you can build the gem and install it using:

- Build your plugin gem
```sh
gem build logstash-filter-awesome.gemspec
```
- Install the plugin from the Logstash home
```sh
# Logstash 2.3 and higher
bin/logstash-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Start Logstash and proceed to test the plugin

## Contributing

All contributions are welcome: ideas, patches, documentation, bug reports, complaints, and even something you drew up on a napkin.

Programming is not a required skill. Whatever you've seen about open source and maintainers or community members  saying "send patches or die" - you will not see that here.

It is more important to the community that you are able to contribute.

For more information about contributing, see the [CONTRIBUTING](https://github.com/elastic/logstash/blob/master/CONTRIBUTING.md) file.

CHANGELOG.md
## 3.3.2
  - Fix an issue that prevented timely shutdown when subscribed to an inactive queue [#65](https://github.com/logstash-plugins/logstash-input-sqs/pull/65)

## 3.3.1
  - Refactoring: used logstash-mixin-aws to leverage shared code to manage `additional_settings` [#64](https://github.com/logstash-plugins/logstash-input-sqs/pull/64)

## 3.3.0
  - Feature: Add `additional_settings` option to fine-grain configuration of AWS client [#61](https://github.com/logstash-plugins/logstash-input-sqs/pull/61)

## 3.2.0
  - Feature: Add `queue_owner_aws_account_id` parameter for cross-account queues [#60](https://github.com/logstash-plugins/logstash-input-sqs/pull/60)

## 3.1.3
  - Fix: retry networking errors (with backoff) [#57](https://github.com/logstash-plugins/logstash-input-sqs/pull/57)

## 3.1.2
  - Added support for multiple events inside same message from SQS [#48](https://github.com/logstash-plugins/logstash-input-sqs/pull/48/files) 

## 3.1.1
  - Docs: Set the default_codec doc attribute.

## 3.1.0
  - Add documentation for endpoint, role_arn and role_session_name #46
  - Fix sample IAM policy to match to match the documentation #32

## 3.0.6
  - Update gemspec summary

## 3.0.5
  - Fix some documentation issues

## 3.0.3
  - Monkey-patch the AWS-SDK to prevent "uninitialized constant" errors.

## 3.0.2
  - Relax constraint on logstash-core-plugin-api to >= 1.60 <= 2.99

## 3.0.1
  - Republish all the gems under jruby.
## 3.0.0
  - Update the plugin to the version 2.0 of the plugin api, this change is required for Logstash 5.0 compatibility. See https://github.com/elastic/logstash/issues/5141
# 2.0.5
  - Depend on logstash-core-plugin-api instead of logstash-core, removing the need to mass update plugins on major releases of logstash
# 2.0.4
  - New dependency requirements for logstash-core for the 5.0 release
## 2.0.3
 - Fixes #22, wrong key use on the stats object
## 2.0.0
 - Plugins were updated to follow the new shutdown semantic, this mainly allows Logstash to instruct input plugins to terminate gracefully, 
   instead of using Thread.raise on the plugins' threads. Ref: https://github.com/elastic/logstash/pull/3895
 - Dependency on logstash-core update to 2.0

# 1.1.0
- AWS ruby SDK v2 upgrade
- Replaces aws-sdk dependencies with mixin-aws
- Removes unnecessary de-allocation
- Move the code into smaller methods to allow easier mocking and testing
- Add the option to configure polling frequency
- Adding a monkey patch to make sure `LogStash::ShutdownSignal` doesn't get catch by AWS RetryError.


spec/inputs/sqs_spec.rb
# encoding: utf-8
require "spec_helper"
require "logstash/devutils/rspec/shared_examples"
require "logstash/inputs/sqs"
require "logstash/errors"
require "logstash/event"
require "logstash/json"
require "aws-sdk"
require "ostruct"

describe LogStash::Inputs::SQS do
  let(:queue_name) { "the-infinite-pandora-box" }
  let(:queue_url) { "https://sqs.test.local/#{queue_name}" }
  let(:config) do
    {
      "region" => "us-east-1",
      "access_key_id" => "123",
      "secret_access_key" => "secret",
      "queue" => queue_name
    }
  end

  let(:input) { LogStash::Inputs::SQS.new(config) }
  let(:decoded_message) { { "bonjour" => "awesome" }  }
  let(:encoded_message)  { double("sqs_message", :body => LogStash::Json::dump(decoded_message)) }

  subject { input }

  let(:mock_sqs) { Aws::SQS::Client.new({ :stub_responses => true }) }


  context "with invalid credentials" do
    before do
      expect(Aws::SQS::Client).to receive(:new).and_return(mock_sqs)
      expect(mock_sqs).to receive(:get_queue_url).with({ :queue_name => queue_name }) { raise Aws::SQS::Errors::ServiceError.new("bad-something", "bad token") }
    end

    it "raises a Configuration error if the credentials are bad" do
      expect { subject.register }.to raise_error(LogStash::ConfigurationError)
    end
  end

  context "valid credentials" do
    let(:queue) { [] }

    it "doesn't raise an error with valid credentials" do
      expect(Aws::SQS::Client).to receive(:new).and_return(mock_sqs)
      expect(mock_sqs).to receive(:get_queue_url).with({ :queue_name => queue_name }).and_return({:queue_url => queue_url })
      expect { subject.register }.not_to raise_error
    end

    context "when queue_aws_account_id option is specified" do
      let(:queue_account_id) { "123456789012" }
      let(:config) do
        {
          "region" => "us-east-1",
          "access_key_id" => "123",
          "secret_access_key" => "secret",
          "queue" => queue_name,
          "queue_owner_aws_account_id" => queue_account_id
        }
      end
      it "passes the option to sqs client" do
        expect(Aws::SQS::Client).to receive(:new).and_return(mock_sqs)
        expect(mock_sqs).to receive(:get_queue_url).with({ :queue_name => queue_name, :queue_owner_aws_account_id => queue_account_id }).and_return({:queue_url => queue_url })
        expect { subject.register }.not_to raise_error
      end
    end

    describe "additional_settings" do
      context "supported settings" do
        let(:config) {
          {
            "additional_settings" => { "force_path_style" => 'true', "ssl_verify_peer" => 'false', "profile" => 'logstash' },
            "queue" => queue_name
          }
        }

        it 'should instantiate Aws::SQS clients with force_path_style set' do
          expect(Aws::SQS::Client).to receive(:new).and_return(mock_sqs)
          # mock a remote call to retrieve the queue URL
          expect(mock_sqs).to receive(:get_queue_url).with({ :queue_name => queue_name }).and_return({:queue_url => queue_url })

          expect(subject.aws_options_hash).to include({:force_path_style => true, :ssl_verify_peer => false, :profile => 'logstash'})

          expect { subject.register }.not_to raise_error
        end
      end

      context "unsupported settings" do
        let(:config) {
          {
            "additional_settings" => { "stub_responses" => 'true', "invalid_option" => "invalid" },
            "queue" => queue_name
          }
        }

        it 'must fail with ArgumentError' do
          expect {subject.register}.to raise_error(ArgumentError, /invalid_option/)
        end
      end

    end

    context "when interrupting the plugin" do
      before do
        expect(Aws::SQS::Client).to receive(:new).and_return(mock_sqs)
        expect(mock_sqs).to receive(:get_queue_url).with({ :queue_name => queue_name }).and_return({:queue_url => queue_url })
        expect(subject).to receive(:poller).and_return(mock_sqs).at_least(:once)

        # We have to make sure we create a bunch of events
        # so we actually really try to stop the plugin.
        # 
        # rspec's `and_yield` allow you to define a fix amount of possible 
        # yielded values and doesn't allow you to create infinite loop.
        # And since we are actually creating thread we need to make sure
        # we have enough work to keep the thread working until we kill it..
        #
        # I haven't found a way to make it rspec friendly
        mock_sqs.instance_eval do
          def poll(polling_options = {})
            loop do
              yield [OpenStruct.new(:body => LogStash::Json::dump({ "message" => "hello world"}))], OpenStruct.new
            end
          end
        end
      end

      it_behaves_like "an interruptible input plugin"
    end

    context "enrich event" do
      let(:event) { LogStash::Event.new }

      let(:message_id) { "123" }
      let(:md5_of_body) { "dr strange" }
      let(:sent_timestamp) { LogStash::Timestamp.new }
      let(:epoch_timestamp) { (sent_timestamp.utc.to_f * 1000).to_i }

      let(:id_field) { "my_id_field" }
      let(:md5_field) { "my_md5_field" }
      let(:sent_timestamp_field) { "my_sent_timestamp_field" }

      let(:message) do
        double("message", :message_id => message_id, :md5_of_body => md5_of_body, :attributes => { LogStash::Inputs::SQS::SENT_TIMESTAMP  => epoch_timestamp } )
      end 

      subject { input.add_sqs_data(event, message) }

      context "when the option is specified" do
        let(:config) do
          {
            "region" => "us-east-1",
            "access_key_id" => "123",
            "secret_access_key" => "secret",
            "queue" => queue_name,
            "id_field" => id_field,
            "md5_field" => md5_field,
            "sent_timestamp_field" => sent_timestamp_field
          }
        end

        it "add the `message_id`" do
          expect(subject.get(id_field)).to eq(message_id)
        end

        it "add the `md5_of_body`" do
          expect(subject.get(md5_field)).to eq(md5_of_body)
        end

        it "add the `sent_timestamp`" do
          expect(subject.get(sent_timestamp_field).to_i).to eq(sent_timestamp.to_i)
        end
      end

      context "when the option isn't specified" do
        it "doesnt add the `message_id`" do
          expect(subject).not_to include(id_field)
        end

        it "doesnt add the `md5_of_body`" do
          expect(subject).not_to include(md5_field)
        end

        it "doesnt add the `sent_timestamp`" do
          expect(subject).not_to include(sent_timestamp_field)
        end
      end
    end

    context "when decoding body" do
      subject { LogStash::Inputs::SQS::new(config.merge({ "codec" => "json" })) }

      it "uses the specified codec" do
        subject.handle_message(encoded_message, queue)
        expect(queue.pop.get("bonjour")).to eq(decoded_message["bonjour"])
      end
    end

    context "receiving messages" do
      before do 
        expect(subject).to receive(:poller).and_return(mock_sqs).at_least(:once)
      end

      it "creates logstash event" do
        expect(mock_sqs).to receive(:poll).with(anything()).and_yield([encoded_message], double("stats"))
        subject.run(queue)
        expect(queue.pop.get("bonjour")).to eq(decoded_message["bonjour"])
      end

      context 'can create multiple events' do
        require "logstash/codecs/json_lines"
        let(:config) { super().merge({ "codec" => "json_lines" }) }
        let(:first_message) { { "sequence" => "first" } }
        let(:second_message) { { "sequence" => "second" } }
        let(:encoded_message)  { double("sqs_message", :body => "#{LogStash::Json::dump(first_message)}\n#{LogStash::Json::dump(second_message)}\n") }

        it 'creates multiple events' do
          expect(mock_sqs).to receive(:poll).with(anything()).and_yield([encoded_message], double("stats"))
          subject.run(queue)
          events = queue.map{ |e|e.get('sequence')}
          expect(events).to match_array([first_message['sequence'], second_message['sequence']])
        end
      end
    end

    context "on errors" do
      let(:payload) { "Hello world" }

      before do
        expect(subject).to receive(:poller).and_return(mock_sqs).at_least(:once)
      end

      context "SQS error" do
        it "retry to fetch messages" do
          # change the poller implementation to raise SQS errors.
          had_error = false
          
          # actually using the child of `Object` to do an expectation of `#sleep`
          expect(subject).to receive(:sleep).with(LogStash::Inputs::SQS::BACKOFF_SLEEP_TIME)
          expect(mock_sqs).to receive(:poll).with(anything()).at_most(2) do
            unless had_error
              had_error = true
              raise Aws::SQS::Errors::ServiceError.new("testing", "testing exception")
            end

            queue << payload
          end

          subject.run(queue)

          expect(queue.size).to eq(1)
          expect(queue.pop).to eq(payload)
        end
      end

      context "SQS error (retries)" do

        it "retry to fetch messages" do
          sleep_time = LogStash::Inputs::SQS::BACKOFF_SLEEP_TIME
          expect(subject).to receive(:sleep).with(sleep_time)
          expect(subject).to receive(:sleep).with(sleep_time * 2)
          expect(subject).to receive(:sleep).with(sleep_time * 4)

          error_count = 0
          expect(mock_sqs).to receive(:poll).with(anything()).at_most(4) do
            error_count += 1
            if error_count <= 3
              raise Aws::SQS::Errors::QueueDoesNotExist.new("testing", "testing exception (#{error_count})")
            end

            queue << payload
          end

          subject.run(queue)

          expect(queue.size).to eq(1)
          expect(queue.pop).to eq(payload)
        end

      end

      context "networking error" do

        before(:all) { require 'seahorse/client/networking_error' }

        it "retry to fetch messages" do
          sleep_time = LogStash::Inputs::SQS::BACKOFF_SLEEP_TIME
          expect(subject).to receive(:sleep).with(sleep_time).twice

          error_count = 0
          expect(mock_sqs).to receive(:poll).with(anything()).at_most(5) do
            error_count += 1
            if error_count == 1
              raise Seahorse::Client::NetworkingError.new(Net::OpenTimeout.new, 'timeout')
            end
            if error_count == 3
              raise Seahorse::Client::NetworkingError.new(SocketError.new('spec-error'))
            end

            queue << payload
          end

          subject.run(queue)
          subject.run(queue)

          expect(queue.size).to eq(2)
          expect(queue.pop).to eq(payload)
        end

      end

      context "other error" do
        it "stops executing the code and raise the exception" do
          expect(mock_sqs).to receive(:poll).with(anything()).at_most(2) do
            raise RuntimeError
          end

          expect { subject.run(queue) }.to raise_error(RuntimeError)
        end
      end
    end
  end
end


spec/support/helpers.rb
# encoding: utf-8
def push_sqs_event(message)
  client = Aws::SQS::Client.new
  queue_url = client.get_queue_url(:queue_name => ENV["SQS_QUEUE_NAME"])

  client.send_message({
    queue_url: queue_url.queue_url,
    message_body: message,
  })
end


spec/integration/sqs_spec.rb
# encoding: utf-8
require "spec_helper"
require "logstash/inputs/sqs"
require "logstash/event"
require "logstash/json"
require "aws-sdk"
require_relative "../support/helpers"
require "thread"

Thread.abort_on_exception = true

describe "LogStash::Inputs::SQS integration", :integration => true do
  let(:decoded_message) { { "drstrange" => "is-he-really-that-strange" } }
  let(:encoded_message) { LogStash::Json.dump(decoded_message) }
  let(:queue) { Queue.new }

  let(:input) { LogStash::Inputs::SQS.new(options) }

  context "with invalid credentials" do
    let(:options) do
      {
        "queue" => "do-not-exist",
        "access_key_id" => "bad_access",
        "secret_access_key" => "bad_secret_key",
        "region" => ENV["AWS_REGION"]
      }
    end

    subject { input }

    it "raises a Configuration error if the credentials are bad" do
      expect { subject.register }.to raise_error(LogStash::ConfigurationError)
    end
  end

  context "with valid credentials" do
    let(:options) do
      {
        "queue" => ENV["SQS_QUEUE_NAME"],
        "access_key_id" => ENV['AWS_ACCESS_KEY_ID'],
        "secret_access_key" => ENV['AWS_SECRET_ACCESS_KEY'],
        "region" => ENV["AWS_REGION"]
      }
    end

    before :each do
      push_sqs_event(encoded_message)
      input.register
      @server = Thread.new { input.run(queue) }
    end

    after do
      @server.kill
    end

    subject { queue.pop }

    it "creates logstash events" do
      expect(subject["drstrange"]).to eq(decoded_message["drstrange"])
    end

    context "when the optionals fields are not specified" do
      let(:id_field) { "my_id_field" }
      let(:md5_field) { "my_md5_field" }
      let(:sent_timestamp_field) { "my_sent_timestamp_field" }

      it "add the `message_id`" do
        expect(subject[id_field]).to be_nil
      end

      it "add the `md5_of_body`" do
        expect(subject[md5_field]).to be_nil
      end

      it "add the `sent_timestamp`" do
        expect(subject[sent_timestamp_field]).to be_nil
      end

    end

    context "when the optionals fields are specified" do
      let(:id_field) { "my_id_field" }
      let(:md5_field) { "my_md5_field" }
      let(:sent_timestamp_field) { "my_sent_timestamp_field" }

      let(:options) do
        {
          "queue" => ENV["SQS_QUEUE_NAME"],
          "access_key_id" => ENV['AWS_ACCESS_KEY_ID'],
          "secret_access_key" => ENV['AWS_SECRET_ACCESS_KEY'],
          "region" => ENV["AWS_REGION"],
          "id_field" => id_field,
          "md5_field" => md5_field,
          "sent_timestamp_field" => sent_timestamp_field
        }
      end

      it "add the `message_id`" do
        expect(subject[id_field]).not_to be_nil
      end

      it "add the `md5_of_body`" do
        expect(subject[md5_field]).not_to be_nil
      end

      it "add the `sent_timestamp`" do
        expect(subject[sent_timestamp_field]).not_to be_nil
      end
    end
  end
end


lib/logstash/inputs/sqs.rb
# encoding: utf-8
#
require "logstash/inputs/threadable"
require "logstash/namespace"
require "logstash/timestamp"
require "logstash/plugin_mixins/aws_config"
require "logstash/errors"
require 'logstash/inputs/sqs/patch'

# Forcibly load all modules marked to be lazily loaded.
#
# It is recommended that this is called prior to launching threads. See
# https://aws.amazon.com/blogs/developer/threading-with-the-aws-sdk-for-ruby/.
Aws.eager_autoload!

# Pull events from an Amazon Web Services Simple Queue Service (SQS) queue.
#
# SQS is a simple, scalable queue system that is part of the
# Amazon Web Services suite of tools.
#
# Although SQS is similar to other queuing systems like AMQP, it
# uses a custom API and requires that you have an AWS account.
# See http://aws.amazon.com/sqs/ for more details on how SQS works,
# what the pricing schedule looks like and how to setup a queue.
#
# To use this plugin, you *must*:
#
#  * Have an AWS account
#  * Setup an SQS queue
#  * Create an identify that has access to consume messages from the queue.
#
# The "consumer" identity must have the following permissions on the queue:
#
#  * `sqs:ChangeMessageVisibility`
#  * `sqs:ChangeMessageVisibilityBatch`
#  * `sqs:DeleteMessage`
#  * `sqs:DeleteMessageBatch`
#  * `sqs:GetQueueAttributes`
#  * `sqs:GetQueueUrl`
#  * `sqs:ListQueues`
#  * `sqs:ReceiveMessage`
#
# Typically, you should setup an IAM policy, create a user and apply the IAM policy to the user.
# A sample policy is as follows:
# [source,json]
#     {
#       "Statement": [
#         {
#           "Action": [
#             "sqs:ChangeMessageVisibility",
#             "sqs:ChangeMessageVisibilityBatch",
#             "sqs:GetQueueAttributes",
#             "sqs:GetQueueUrl",
#             "sqs:ListQueues",
#             "sqs:SendMessage",
#             "sqs:SendMessageBatch"
#           ],
#           "Effect": "Allow",
#           "Resource": [
#             "arn:aws:sqs:us-east-1:123456789012:Logstash"
#           ]
#         }
#       ]
#     }
#
# See http://aws.amazon.com/iam/ for more details on setting up AWS identities.
#
class LogStash::Inputs::SQS < LogStash::Inputs::Threadable
  include LogStash::PluginMixins::AwsConfig::V2

  MAX_TIME_BEFORE_GIVING_UP = 60
  MAX_MESSAGES_TO_FETCH = 10 # Between 1-10 in the AWS-SDK doc
  SENT_TIMESTAMP = "SentTimestamp"
  SQS_ATTRIBUTES = [SENT_TIMESTAMP]
  BACKOFF_SLEEP_TIME = 1
  BACKOFF_FACTOR = 2
  DEFAULT_POLLING_FREQUENCY = 20

  config_name "sqs"

  default :codec, "json"

  config :additional_settings, :validate => :hash, :default => {}

  # Name of the SQS Queue name to pull messages from. Note that this is just the name of the queue, not the URL or ARN.
  config :queue, :validate => :string, :required => true

  # Account ID of the AWS account which owns the queue.
  config :queue_owner_aws_account_id, :validate => :string, :required => false

  # Name of the event field in which to store the SQS message ID
  config :id_field, :validate => :string

  # Name of the event field in which to store the SQS message MD5 checksum
  config :md5_field, :validate => :string

  # Name of the event field in which to store the SQS message Sent Timestamp
  config :sent_timestamp_field, :validate => :string

  # Polling frequency, default is 20 seconds
  config :polling_frequency, :validate => :number, :default => DEFAULT_POLLING_FREQUENCY

  attr_reader :poller

  def register
    require "aws-sdk"
    @logger.info("Registering SQS input", :queue => @queue, :queue_owner_aws_account_id => @queue_owner_aws_account_id)

    setup_queue
  end

  def queue_url(aws_sqs_client)
    if @queue_owner_aws_account_id
      return aws_sqs_client.get_queue_url({:queue_name => @queue, :queue_owner_aws_account_id => @queue_owner_aws_account_id})[:queue_url]
    else
      return aws_sqs_client.get_queue_url(:queue_name => @queue)[:queue_url]
    end
  end

  def setup_queue
    aws_sqs_client = Aws::SQS::Client.new(aws_options_hash || {})
    poller = Aws::SQS::QueuePoller.new(queue_url(aws_sqs_client), :client => aws_sqs_client)
    poller.before_request { |stats| throw :stop_polling if stop? }

    @poller = poller
  rescue Aws::SQS::Errors::ServiceError, Seahorse::Client::NetworkingError => e
    @logger.error("Cannot establish connection to Amazon SQS", exception_details(e))
    raise LogStash::ConfigurationError, "Verify the SQS queue name and your credentials"
  end

  def polling_options
    { 
      :max_number_of_messages => MAX_MESSAGES_TO_FETCH,
      :attribute_names => SQS_ATTRIBUTES,
      :wait_time_seconds => @polling_frequency
    }
  end

  def add_sqs_data(event, message)
    event.set(@id_field, message.message_id) if @id_field
    event.set(@md5_field, message.md5_of_body) if @md5_field
    event.set(@sent_timestamp_field, convert_epoch_to_timestamp(message.attributes[SENT_TIMESTAMP])) if @sent_timestamp_field
    event
  end

  def handle_message(message, output_queue)
    @codec.decode(message.body) do |event|
      add_sqs_data(event, message)
      decorate(event)
      output_queue << event
    end
  end

  def run(output_queue)
    @logger.debug("Polling SQS queue", :polling_options => polling_options)

    run_with_backoff do
      poller.poll(polling_options) do |messages, stats|
        break if stop?
        messages.each {|message| handle_message(message, output_queue) }
        @logger.debug("SQS Stats:", :request_count => stats.request_count,
                      :received_message_count => stats.received_message_count,
                      :last_message_received_at => stats.last_message_received_at) if @logger.debug?
      end
    end
  end

  private

  # Runs an AWS request inside a Ruby block with an exponential backoff in case
  # we experience a ServiceError.
  #
  # @param [Block] block Ruby code block to execute.
  def run_with_backoff(&block)
    sleep_time = BACKOFF_SLEEP_TIME
    begin
      block.call
    rescue Aws::SQS::Errors::ServiceError, Seahorse::Client::NetworkingError => e
      @logger.warn("SQS error ... retrying with exponential backoff", exception_details(e, sleep_time))
      sleep_time = backoff_sleep(sleep_time)
      retry
    end
  end

  def backoff_sleep(sleep_time)
    sleep(sleep_time)
    sleep_time > MAX_TIME_BEFORE_GIVING_UP ? sleep_time : sleep_time * BACKOFF_FACTOR
  end

  def convert_epoch_to_timestamp(time)
    LogStash::Timestamp.at(time.to_i / 1000)
  end

  def exception_details(e, sleep_time = nil)
    details = { :queue => @queue, :exception => e.class, :message => e.message }
    details[:code] = e.code if e.is_a?(Aws::SQS::Errors::ServiceError) && e.code
    details[:cause] = e.original_error if e.respond_to?(:original_error) && e.original_error # Seahorse::Client::NetworkingError
    details[:sleep_time] = sleep_time if sleep_time
    details[:backtrace] = e.backtrace if @logger.debug?
    details
  end

end # class LogStash::Inputs::SQS
